zoomable-image/coil3/src/main/kotlin/me/saket/telephoto/zoomable/coil3/Coil3ImageSource.kt

```diff
@@ -0,0 +1,242 @@
+        sizeResolver = { canvasSize.first().toCoilSize() }
+      )
+    }
+    return resolver.resolved
+  }
+
+  private fun Size.toCoilSize() = CoilSize(
+    width = if (width.isFinite()) Dimension(width.roundToInt()) else Dimension.Undefined,
+    height = if (height.isFinite()) Dimension(height.roundToInt()) else Dimension.Undefined
+  )
+}
+
+internal class Resolver(
+  internal val request: ImageRequest,
+  internal val imageLoader: ImageLoader,
+  private val sizeResolver: SizeResolver,
+) : RememberWorker() {
+
+  internal var resolved: ResolveResult by mutableStateOf(
+    ResolveResult(delegate = null)
+  )
+
+  override suspend fun work() {
+    work(skipMemoryCache = false)
+  }
+
+  private suspend fun work(skipMemoryCache: Boolean) {
+    val imageLoader: ImageLoader = imageLoader
+      .newBuilder()
+      // Ignore "no-store" http headers if they're present and always cache images to disk. Otherwise,
+      // telephoto will be unable to sub-sample large images directly from coil's memory cache.
+      //.respectCacheHeaders(false) // todo: how do I do this with coil3?
+      .build()
+
+    val result = imageLoader.execute(
+      request.newBuilder()
+        .size(request.defined.sizeResolver ?: sizeResolver)
+        // There's no easy way to be certain whether an image will require sub-sampling in
+        // advance so assume it'll be needed and force Coil to write this image to disk.
+        .diskCachePolicy(
+          when (request.diskCachePolicy) {
+            CachePolicy.ENABLED -> CachePolicy.ENABLED
+            CachePolicy.READ_ONLY -> CachePolicy.ENABLED
+            CachePolicy.WRITE_ONLY,
+            CachePolicy.DISABLED -> CachePolicy.WRITE_ONLY
+          }
+        )
+        .memoryCachePolicy(
+          if (skipMemoryCache) CachePolicy.WRITE_ONLY else request.memoryCachePolicy
+        )
+        // This will unfortunately replace any existing target, but it is also the only
+        // way to read placeholder images set using ImageRequest#placeholderMemoryCacheKey.
+        // Placeholder images should be small in size so sub-sampling isn't needed here.
+        .target(
+          onStart = {
+            resolved = resolved.copy(
+              placeholder = it?.asPainter(request.context),
+            )
+          }
+        )
+        // Increase memory cache hit rate because the image will anyway fit the canvas
+        // size at draw time.
+        .precision(
+          when (request.defined.precision) {
+            Precision.EXACT -> request.precision
+            else -> Precision.INEXACT
+          }
+        )
+        // todo: should coil spend any time doing this?
+        //  counter-arg: i might want to enable sub-sampling only if the bitmap size exceeds the display size.
+        //  counter-counter-arg: if CacheControlCacheStrategy is used and the image isn't saved to disk, it's better to display a smaller image without sub-sampling.
+        .maxBitmapSize(CoilSize.ORIGINAL)
```

Comment:
@colinrtwhite some questions for you:

a) I'm leaning towards allowing coil to downsize large images to prevent OOMs, especially if the images are never written to the disk and can't be sub-sampled. Is there a way to read the original image size? That'll allow me to adjust my sub-sampling strategy in the future to enable it only for large images rather than all bitmaps.

b) If I decide against `(a)`, is any time saved when `Size.ORIGINAL` is used? Or does coil always resize images even if the before and after sizes are the same?

Reply:
@saket `maxBitmapSize` is mostly a guard to prevent creating very large bitmaps - I would recommend leaving it as the default (solution A). For instance, if you have a very long image (e.g. 100x1,000) and try to `Scale.FILL` it into a big target (e.g. 1000x1,000); that will create a 1,000x10,000 bitmap and crash.

If you set it to `maxBitmapSize(Size.ORIGINAL)` you probably won't save any time. In fact, I believe it's likely faster to load a subsampled version of the image (especially if it's smaller by powers of 2). Though, I'm not 100% sure.

There's [no API](https://github.com/coil-kt/coil/issues/1213) to read the image's original size unfortunately. I've been a bit hesitant to add this for two main reasons:

- Some images don't have clear image dimensions (mostly SVGs) and the way Coil determines its source dimensions is a bit of an implementation detail.
- I want to avoid exposing image metadata in Coil as it can get complicated quickly.

Reply:
Fair points, thanks!

> I would recommend leaving it as the default (solution A). For instance, if you have a very long image (e.g. 100x1,000) and try to Scale.FILL it into a big target (e.g. 1000x1,000); that will create a 1,000x10,000 bitmap and crash.

Won't this only happen if I'm actually displaying the bitmap in an image composable? Telephoto is only reading the downloaded file path from the disk cache and then handling the rendering part using `SubSamplingImage()`.